File: easy/144_Binary_Tree_Preorder_Traversal.py
# Given the root of a binary tree, return the preorder traversal of its nodes' value
#
# Input: root = [1,null,2,3]
# Output: [1,2,3]
#
# Input: root = [1]
# Output: [1]
from typing import List, Optional


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def preorderTraversal(self, root: Optional[TreeNode]) -> List[int]:
        res = list()
        if root:
            res.append(root.val)
            res += self.preorderTraversal(root.left)
            res += self.preorderTraversal(root.right)
        return res

    def preorderTraversal2(self, root: Optional[TreeNode]) -> List[int]:
        '''
        Space comp O(N)
        Time comp O(N)
        :param root:
        :return:
        '''
        res = list()
        queue = [root]
        while queue:
            current_node = queue.pop()
            res.append(current_node.val)
            if current_node.right is not None:
                queue.append(current_node.right)
            if current_node.left is not None:
                queue.append(current_node.left)
        return res

File: go/easy/144_Binary_Tree_Preorder_Traversal.go
package main

import "fmt"

//Given the root of a binary tree, return the preorder traversal of its nodes' values.

//Example 1:
//Input: root = [1,null,2,3]
//Output: [1,2,3]

//Example 2:
//Input: root = []
//Output: []

//Example 3:
//Input: root = [1]
//Output: [1]

//type TreeNode struct {
//	Val   int
//	Left  *TreeNode
//	Right *TreeNode
//}

func preorderHelper(root *TreeNode, res *[]int) {
	if root == nil {
		return
	}
	*res = append(*res, root.Val)
	preorderHelper(root.Left, res)
	preorderHelper(root.Right, res)
}

func preorderTraversal(root *TreeNode) []int {
	var res []int
	preorderHelper(root, &res)
	return res
}

func main() {
	tree := &TreeNode{
		Val: 1,
		Left: &TreeNode{
			Val: 0,
		},
		Right: &TreeNode{
			Val: 2,
			Left: &TreeNode{
				Val: 3,
			},
		},
	}
	fmt.Println(preorderTraversal(tree))
}
